File: src/audioWfc/constraints/MelodyStartsOnNoteHardConstraint.java
package audioWfc.constraints;

import audioWfc.HigherValues;
import audioWfc.OctavedNote;
import audioWfc.Tile;
import audioWfc.constraints.grabbers.Grabber;
import audioWfc.musicTheory.Note;

public class MelodyStartsOnNoteHardConstraint implements HardConstraint<OctavedNote>{
    private Grabber<Note> grabber;

    public MelodyStartsOnNoteHardConstraint(Grabber<Note> grabber){
        this.grabber = grabber;
    }

    @Override
    public boolean check(Tile<OctavedNote> tile, HigherValues higherValues) {
        if(tile.getPosition() != 0) return true;
        return tile.getValue().getNote().equals(grabber.grab(higherValues));
    }

}

File: vite-project/src/wfc/constraints/MelodyStartsOnNoteHardConstraint.ts
import { OctavedNote, Note } from "../../music_theory/Note"
import { Grabber } from "../Grabber"
import { HigherValues } from "../HigherValues"
import { Tile } from "../Tile"
import { HardConstraint } from "./concepts/Constraint"
import { noteConstraintTypeToName } from "./constraintUtils"

export class MelodyStartsOnNoteHardConstraint implements HardConstraint<OctavedNote> {
	private grabber: Grabber<Note>
	name = noteConstraintTypeToName.get("MelodyStartsOnNoteHardConstraint") as string
	configText = () => `Note: ${this.grabber.configText()}`
	constructor(grabber: Grabber<Note>) {
		this.grabber = grabber
	}

	check(tile: Tile<OctavedNote>, higherValues: HigherValues): boolean {
		if(tile.getPosition() != 0) return true
		const note = tile.getValue()
		const startNote = this.grabber.grab(higherValues)
		return note.getNote() == startNote
	}
}
